swap/src/asb/event_loop.rs

```diff
@@ -318,16 +318,30 @@ where
         &mut self,
         min_buy: bitcoin::Amount,
         max_buy: bitcoin::Amount,
+        monero_wallet: Arc<monero::Wallet>,
     ) -> Result<BidQuote> {
+        let max_bitcoin_for_monero: bitcoin::Amount;
+        let mut max_bitcoin_buy = max_buy;
+        let monero_balance = monero_wallet.get_balance().await?.as_monero();
+
         let rate = self
             .latest_rate
             .latest_rate()
             .context("Failed to get latest rate")?;
 
+        max_bitcoin_for_monero = bitcoin::Amount::from_btc(monero_balance * rate.ask()?.as_btc())?;
```

Comment:
No floating point math please. Either operate on sats and piconeros (which are both u64 I think) or use `Decimal` if you can't use natural numbers.

Reply:
Also, it is a bit of an anti-pattern to extract the raw value from a newtype (the `as_monero` function call) because the type system no longer helps you in correctly interpreting the result of the computation.

You should be able to express this functionality by implementing `Mul` for `Rate` and `Monero`.

Reply:
Thank you for the input! I ended up using decimals at the end. I think the UX for the BTC and XMR nomenclature is better than Piconero and Satoshi one